### PR TITLE
feat(merkle-distributor): Gas optimizations for claimMulti with precomputations

### PR DESCRIPTION
**Pull request by nicholaspai**
**Motivation**

Alternative implementation for #2672 in which the caller to `claimMulti` can pass in an `address[]` of `rewardTokens` and an `address[]` of accounts to claim for, allowing the smart contract to reduce the amount of Transfers submitted for a batch claim. Alternative implementation to #2672 which reduces gas costs but increases code complexity.

**Details**

This may be unnecessary for some PRs. Catch-all for detailed explanations about the implementation decisions and implications of the change.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

**Issue(s)**

If merged, closes #2672

### CODE REVIEW

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 1]
```diff
@@ -206,21 +218,21 @@ contract MerkleDistributor is Ownable {
...
-        _setClaimed(claim.windowIndex, claim.accountIndex);
+        _setClaimed(_claim.windowIndex, _claim.accountIndex);
         emit Claimed(
             msg.sender,
-            claim.windowIndex,
-            claim.account,
-            claim.accountIndex,
-            claim.amount,
-            address(merkleWindows[claim.windowIndex].rewardToken)
+            _claim.windowIndex,
```
Comment (nicholaspai) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:231]: ALl of these `claim` -> `_claim` renamings are to eliminate compiler warnings about variable shadowing

File: packages/core/test/merkle-distributor/MerkleDistributor.js [hunk 2]
```diff
@@ -442,102 +456,95 @@ contract("MerkleDistributor.js", function(accounts) {
...
-            merkleProof: merkleTrees[0].getProof(leaf2.leaf)
-          }
-        ];
-
-        // Show that verifyProof works for claims individually.
-        for (let claim of invalidClaims) {
-          assert.isTrue(await merkleDistributor.verifyClaim(claim));
+      it("gas", async function() {
+        const txn = await merkleDistributor.claimMulti(batchedClaims, rewardRecipientAddresses, rewardTokenAddresses);
+        assert.equal(txn.receipt.gasUsed, 1145721);
```
Comment (nicholaspai) [inline packages/core/test/merkle-distributor/MerkleDistributor.js:509]: Note that `claimMulti` on this trivial tree actually introduces a 40,000 gas increase compared to #2672, but on the larger 100k leaf tree we see gas savings for this implementation.

File: packages/core/test/merkle-distributor/MerkleDistributor.js [hunk 3]
```diff
@@ -615,7 +624,22 @@ contract("MerkleDistributor.js", function(accounts) {
...
+        batchedClaims.push({
+          windowIndex: windowIndex,
+          account: leaf.account,
+          accountIndex: leaf.accountIndex,
+          amount: leaf.amount,
+          merkleProof: proof
+        });
+      }
+      const tx = await merkleDistributor.claimMulti(batchedClaims, [rando], [rewardToken.address]);
+      assert.equal(Math.floor(tx.receipt.gasUsed / batchedClaims.length), 49304);
```
Comment (nicholaspai) [inline packages/core/test/merkle-distributor/MerkleDistributor.js:642]: Compare to 56846 gas for #2672 [here](https://github.com/UMAprotocol/protocol/pull/2672/files#r590803080)

File: packages/core/test/merkle-distributor/MerkleDistributor.js [hunk 4]
```diff
@@ -637,7 +661,22 @@ contract("MerkleDistributor.js", function(accounts) {
...
+        batchedClaims.push({
+          windowIndex: windowIndex,
+          account: leaf.account,
+          accountIndex: leaf.accountIndex,
+          amount: leaf.amount,
+          merkleProof: proof
+        });
+      }
+      const tx = await merkleDistributor.claimMulti(batchedClaims, [rando], [rewardToken.address]);
+      assert.equal(Math.floor(tx.receipt.gasUsed / batchedClaims.length), 30858);
```
Comment (nicholaspai) [inline packages/core/test/merkle-distributor/MerkleDistributor.js:679]: Compare to 38401 for #2672 [here](https://github.com/UMAprotocol/protocol/pull/2672/files#r590803497)

File: packages/core/test/merkle-distributor/MerkleDistributor.js [hunk 5]
```diff
@@ -633,35 +608,148 @@ contract("MerkleDistributor.js", function(accounts) {
...
+            account: leaf.account,
+            accountIndex: leaf.accountIndex,
+            amount: leaf.amount,
+            merkleProof: proof
+          });
+        }
+        const tx = await merkleDistributor.claimMulti(batchedClaims, possibleRecipients, [rewardToken.address]);
+        assert.equal(Math.floor(tx.receipt.gasUsed / batchedClaims.length), 53119);
+      });
+      it("gas amortized one account across multiple windows with different reward tokens", async function() {
```
Comment (nicholaspai) [inline packages/core/test/merkle-distributor/MerkleDistributor.js:695]: Here

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol [hunk 6]
```diff
@@ -127,34 +127,50 @@ contract MerkleDistributor is Ownable {
...
-    // then this function would still reduce to multiple ERC20.transfer calls.
+    // Batch claims for `accounts` for the given `rewardTokens`. Claims for other accounts or
+    // other reward tokens will be skipped. If any claims for the specified `accounts` with a listed
+    // reward token are invalid then the entire method will revert.
     function claimMulti(
         Claim[] memory claims,
-        address rewardToken,
-        address account
+        address[] calldata accounts,
+        address[] calldata rewardTokens
```
Comment (daywiss) [inline packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:136]: i dont understand why these 2 params are required, cant you infer this from the claims array? it also makes this kind of error prone as there arent checks if all the arrays match with all the claims.

**Review by daywiss — COMMENTED** [review]
i think i prefer the simpler version of this where you can just provide the claim array with slightly higher gas usage.

File: packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol (hunk 6 above)
Comment (nicholaspai) [reply packages/core/contracts/merkle-distributor/implementation/MerkleDistributor.sol:136]: So you don't know how many unique `accounts` or `rewardTokens` are in the `claims` array, unless you loop through it once to initialize those arrays. So I passed on the burden to the caller here to supply the accounts and rewardTokens to disburse

**Comment by nicholaspai** [comment]
Closed by #2672, [details](https://github.com/UMAprotocol/protocol/pull/2672#issuecomment-796143682)

**Pull request closed without merging**